<a href="https://colab.research.google.com/github/michiWS1920/ADS20/blob/master/ADS_Programming_Assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Programming Assignment 1
**Data Wrangling**

In this first problem set you will use your data wrangling skills to answer (hopefully) interessting questions regarding the COVID-19 pandemic.


To perform your analysis you have to load the library ```tidyverse```. If you want to use additional libraries you can add them to the following cell:

In [1]:
library(tidyverse)
library(lubridate)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.0     ✔ purrr   0.3.4
✔ tibble  3.0.1     ✔ dplyr   0.8.5
✔ tidyr   1.0.3     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


Attaching package: ‘lubridate’


The following objects are masked from ‘package:dplyr’:

    intersect, setdiff, union


The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union




## Task 1
**COVID-19 cases worldwide**

For you first task you will work with data provided by the [EU Open Data Portal](https://data.europa.eu/euodp/en/home). We will use a dataset providing information on the geographic distribution of COVID-19 cases worldwide. You can download the .csv file [here](https://data.europa.eu/euodp/en/data/dataset/covid-19-coronavirus-data/resource/260bbbde-2316-40eb-aec3-7cd7bfc2f590) or load it directly into your notebook by using the download link https://opendata.ecdc.europa.eu/covid19/casedistribution/csv .

Your first task is to load the data into a dataframe:

In [0]:
# Insert your code here
covid <- as.data.frame(read.csv(url("https://opendata.ecdc.europa.eu/covid19/casedistribution/csv")))
covid

dateRep    day month year cases deaths countriesAndTerritories geoId
1     20/05/2020 20  5     2020  581   5     Afghanistan             AF   
2     19/05/2020 19  5     2020  408   4     Afghanistan             AF   
3     18/05/2020 18  5     2020  262   1     Afghanistan             AF   
4     17/05/2020 17  5     2020    0   0     Afghanistan             AF   
5     16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
6     15/05/2020 15  5     2020  113   6     Afghanistan             AF   
7     14/05/2020 14  5     2020  259   3     Afghanistan             AF   
8     13/05/2020 13  5     2020  280   5     Afghanistan             AF   
9     12/05/2020 12  5     2020  285   2     Afghanistan             AF   
10    11/05/2020 11  5     2020  369   5     Afghanistan             AF   
11    10/05/2020 10  5     2020  255   6     Afghanistan             AF   
12    09/05/2020  9  5     2020  215   3     Afghanistan             AF   
13    08/05/2020  8  5     2020  171   2     Afghanistan             AF   
14    07/05/2020  7  5     2020  168   9     Afghanistan             AF   
15    06/05/2020  6  5     2020  330   5     Afghanistan             AF   
16    05/05/2020  5  5     2020  190   5     Afghanistan             AF   
17    04/05/2020  4  5     2020  235  13     Afghanistan             AF   
18    03/05/2020  3  5     2020  134   4     Afghanistan             AF   
19    02/05/2020  2  5     2020  164   4     Afghanistan             AF   
20    01/05/2020  1  5     2020  222   4     Afghanistan             AF   
21    30/04/2020 30  4     2020  122   0     Afghanistan             AF   
22    29/04/2020 29  4     2020  124   3     Afghanistan             AF   
23    28/04/2020 28  4     2020  172   0     Afghanistan             AF   
24    27/04/2020 27  4     2020   68  10     Afghanistan             AF   
25    26/04/2020 26  4     2020  112   4     Afghanistan             AF   
26    25/04/2020 25  4     2020   70   1     Afghanistan             AF   
27    24/04/2020 24  4     2020  105   2     Afghanistan             AF   
28    23/04/2020 23  4     2020   84   4     Afghanistan             AF   
29    22/04/2020 22  4     2020   61   1     Afghanistan             AF   
30    21/04/2020 21  4     2020   35   2     Afghanistan             AF   
⋮     ⋮          ⋮   ⋮     ⋮    ⋮     ⋮      ⋮                       ⋮    
18174 19/04/2020 19  4     2020 1     0      Zimbabwe                ZW   
18175 18/04/2020 18  4     2020 0     0      Zimbabwe                ZW   
18176 17/04/2020 17  4     2020 1     0      Zimbabwe                ZW   
18177 16/04/2020 16  4     2020 6     0      Zimbabwe                ZW   
18178 15/04/2020 15  4     2020 0     0      Zimbabwe                ZW   
18179 14/04/2020 14  4     2020 3     0      Zimbabwe                ZW   
18180 13/04/2020 13  4     2020 0     0      Zimbabwe                ZW   
18181 12/04/2020 12  4     2020 3     0      Zimbabwe                ZW   
18182 11/04/2020 11  4     2020 0     0      Zimbabwe                ZW   
18183 10/04/2020 10  4     2020 0     1      Zimbabwe                ZW   
18184 09/04/2020  9  4     2020 1     1      Zimbabwe                ZW   
18185 08/04/2020  8  4     2020 1     0      Zimbabwe                ZW   
18186 07/04/2020  7  4     2020 0     0      Zimbabwe                ZW   
18187 06/04/2020  6  4     2020 0     0      Zimbabwe                ZW   
18188 05/04/2020  5  4     2020 0     0      Zimbabwe                ZW   
18189 04/04/2020  4  4     2020 1     0      Zimbabwe                ZW   
18190 03/04/2020  3  4     2020 0     0      Zimbabwe                ZW   
18191 02/04/2020  2  4     2020 0     0      Zimbabwe                ZW   
18192 01/04/2020  1  4     2020 1     0      Zimbabwe                ZW   
18193 31/03/2020 31  3     2020 0     0      Zimbabwe                ZW   
18194 30/03/2020 30  3     2020 0     0      Zimbabwe                ZW   
18195 29/03/2020 29  3     2020

You should always familiarize yourself with the dataset at hand befor you do calculations and try to gain insights.

Use functions such as ```head()```, ```sample_n()```, ```summary()```, etc. to deepen your understanding of the available data and briefly describe the dataset (e.g., number of observations, sorting, variables, ...):

In [0]:
# Insert your code here
covid %>%
head()

dateRep    day month year cases deaths countriesAndTerritories geoId
1 15/05/2020 15  5     2020 113   6      Afghanistan             AF   
2 14/05/2020 14  5     2020 259   3      Afghanistan             AF   
3 13/05/2020 13  5     2020 280   5      Afghanistan             AF   
4 12/05/2020 12  5     2020 285   2      Afghanistan             AF   
5 11/05/2020 11  5     2020 369   5      Afghanistan             AF   
6 10/05/2020 10  5     2020 255   6      Afghanistan             AF   
  countryterritoryCode popData2018 continentExp
1 AFG                  37172386    Asia        
2 AFG                  37172386    Asia        
3 AFG                  37172386    Asia        
4 AFG                  37172386    Asia        
5 AFG                  37172386    Asia        
6 AFG                  37172386    Asia

In [0]:
covid %>%
sample_n(10) %>%
summary()

       dateRep       day           month           year          cases       
 25/04/2020:2   Min.   : 2.0   Min.   :1.00   Min.   :2020   Min.   :  0.00  
 02/05/2020:1   1st Qu.: 5.5   1st Qu.:3.25   1st Qu.:2020   1st Qu.:  2.25  
 04/01/2020:1   Median :15.0   Median :4.00   Median :2020   Median : 23.00  
 05/04/2020:1   Mean   :15.2   Mean   :3.50   Mean   :2020   Mean   : 54.10  
 07/04/2020:1   3rd Qu.:25.0   3rd Qu.:4.00   3rd Qu.:2020   3rd Qu.: 37.75  
 09/04/2020:1   Max.   :29.0   Max.   :5.00   Max.   :2020   Max.   :312.00  
 (Other)   :3                                                                
     deaths     countriesAndTerritories     geoId   countryterritoryCode
 Min.   : 0.0   Algeria:1               AT     :1   AUT    :1           
 1st Qu.: 0.0   Austria:1               BT     :1   BTN    :1           
 Median : 2.5   Bhutan :1               DK     :1   DNK    :1           
 Mean   : 3.5   Denmark:1               DZ     :1   DZA    :1           
 3rd Qu.: 5

In [0]:
covid %>%
distinct(countriesAndTerritories)

countriesAndTerritories          
1   Afghanistan                      
2   Albania                          
3   Algeria                          
4   Andorra                          
5   Angola                           
6   Anguilla                         
7   Antigua_and_Barbuda              
8   Argentina                        
9   Armenia                          
10  Aruba                            
11  Australia                        
12  Austria                          
13  Azerbaijan                       
14  Bahamas                          
15  Bahrain                          
16  Bangladesh                       
17  Barbados                         
18  Belarus                          
19  Belgium                          
20  Belize                           
21  Benin                            
22  Bermuda                          
23  Bhutan                           
24  Bolivia                          
25  Bonaire, Saint Eustatius and Saba
26  Bosnia_and_Herzegovina           
27  Botswana                         
28  Brazil                           
29  British_Virgin_Islands           
30  Brunei_Darussalam                
⋮   ⋮                                
181 Sri_Lanka                        
182 Sudan                            
183 Suriname                         
184 Sweden                           
185 Switzerland                      
186 Syria                            
187 Taiwan                           
188 Tajikistan                       
189 Thailand                         
190 Timor_Leste                      
191 Togo                             
192 Trinidad_and_Tobago              
193 Tunisia                          
194 Turkey                           
195 Turks_and_Caicos_islands         
196 Uganda                           
197 Ukraine                          
198 United_Arab_Emirates             
199 United_Kingdom                   
200 United_Republic_of_Tanzania      
201 United_States_of_America         
202 United_States_Virgin_Islands     
203 Uruguay                          
204 Uzbekistan                       
205 Venezuela                        
206 Vietnam                          
207 Western_Sahara                   
208 Yemen                            
209 Zambia                           
210 Zimbabwe

In [0]:
covid %>%
arrange(desc(dateRep)) %>%
head()

dateRep    day month year cases deaths countriesAndTerritories geoId
1 31/12/2019 31  12    2019 0     0      Afghanistan             AF   
2 31/12/2019 31  12    2019 0     0      Algeria                 DZ   
3 31/12/2019 31  12    2019 0     0      Armenia                 AM   
4 31/12/2019 31  12    2019 0     0      Australia               AU   
5 31/12/2019 31  12    2019 0     0      Austria                 AT   
6 31/12/2019 31  12    2019 0     0      Azerbaijan              AZ   
  countryterritoryCode popData2018 continentExp
1 AFG                  37172386    Asia        
2 DZA                  42228429    Africa      
3 ARM                   2951776    Europe      
4 AUS                  24992369    Oceania     
5 AUT                   8847037    Europe      
6 AZE                   9942334    Europe

In [0]:
covid %>%
arrange(desc(-dateRep)) %>%
sample_n(10)

Warning message in Ops.factor(dateRep):
“‘-’ not meaningful for factors”


dateRep    day month year cases deaths countriesAndTerritories geoId
1  28/03/2020 28  3     2020    1    0    Curaçao                 CW   
2  27/03/2020 27  3     2020    0    0    Tunisia                 TN   
3  02/04/2020  2  4     2020    0    0    Cape_Verde              CV   
4  04/03/2020  4  3     2020    2    0    Bahrain                 BH   
5  29/01/2020 29  1     2020    0    0    Lebanon                 LB   
6  04/04/2020  4  4     2020 6663  932    Spain                   ES   
7  19/01/2020 19  1     2020    0    0    Iraq                    IQ   
8  06/04/2020  6  4     2020  387   28    Sweden                  SE   
9  29/04/2020 29  4     2020    0    0    Benin                   BJ   
10 24/03/2020 24  3     2020   35    0    Argentina               AR   
   countryterritoryCode popData2018 continentExp
1  CUW                    159849    America     
2  TUN                  11565204    Africa      
3  CPV                    543767    Africa      
4  BHR                   1569439    Asia        
5  LBN                   6848925    Asia        
6  ESP                  46723749    Europe      
7  IRQ                  38433600    Asia        
8  SWE                  10183175    Europe      
9  BEN                  11485048    Africa      
10 ARG                  44494502    America

*Describe the dataset here*

o deepen your understanding of the available data and briefly describe the dataset (e.g., number of observations, sorting, variables, ...):

Dataset shows data regarding covid cases. 
* Counts new cases/deaths per day per country
* Shows population data for said country/continent
* 17158 rows of data
* sorted by country and date
* Starting with 31/12/2019 until today (15/5/2020)
* includes 210 different countries
* 11 different variables - date in D/M/Y format. 3 columns for day/month/year split up - Cases - Deaths - Country - GeoID - CountryCode - Population data from 2018 and the Continent


Next, you should filter and sort the dataset to answer the following questions:

1.   Which 5 days had the highest number of reported new cases in Germany?
2.   Which 5 days had the highest number of reported deaths in Brazil?

In [0]:
# Insert your code here
covid %>%
filter(countriesAndTerritories == "Germany") %>%
arrange(desc(cases)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories geoId
1 28/03/2020 28  3     2020 6294   72    Germany                 DE   
2 03/04/2020  3  4     2020 6174  145    Germany                 DE   
3 02/04/2020  2  4     2020 6156  140    Germany                 DE   
4 04/04/2020  4  4     2020 6082  141    Germany                 DE   
5 20/03/2020 20  3     2020 5940   30    Germany                 DE   
  countryterritoryCode popData2018 continentExp
1 DEU                  82927922    Europe      
2 DEU                  82927922    Europe      
3 DEU                  82927922    Europe      
4 DEU                  82927922    Europe      
5 DEU                  82927922    Europe

In [0]:
covid %>%
filter(countriesAndTerritories == "Brazil") %>%
arrange(desc(deaths)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories geoId
1 13/05/2020 13  5     2020  9258 881    Brazil                  BR   
2 15/05/2020 15  5     2020 13944 844    Brazil                  BR   
3 09/05/2020  9  5     2020 10222 751    Brazil                  BR   
4 14/05/2020 14  5     2020 11385 749    Brazil                  BR   
5 10/05/2020 10  5     2020 10611 730    Brazil                  BR   
  countryterritoryCode popData2018 continentExp
1 BRA                  209469333   America     
2 BRA                  209469333   America     
3 BRA                  209469333   America     
4 BRA                  209469333   America     
5 BRA                  209469333   America

In [0]:
covid %>%
filter(countriesAndTerritories == "United_States_of_America") %>%
arrange(desc(cases)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories  geoId
1 26/04/2020 26  4     2020 48529 2172   United_States_of_America US   
2 22/04/2020 22  4     2020 37289 2524   United_States_of_America US   
3 11/04/2020 11  4     2020 35527 2087   United_States_of_America US   
4 05/04/2020  5  4     2020 34272 1344   United_States_of_America US   
5 02/05/2020  2  5     2020 33955 2062   United_States_of_America US   
  countryterritoryCode popData2018 continentExp
1 USA                  327167434   America     
2 USA                  327167434   America     
3 USA                  327167434   America     
4 USA                  327167434   America     
5 USA                  327167434   America

Having established first insights, you now have to aggregate and transform the dataset for further analysis. 

First, calculate the number of total confirmed cases and the number of total deaths for each country. Which countries are hit hardest in terms of total cases and total deaths? 

In [0]:
covid %>%
group_by(countriesAndTerritories) %>%
summarise(totalcases = sum(cases), totaldeaths = sum(deaths)) %>%
arrange(-totalcases) -> covid_sum
covid_sum %>%
head()

countriesAndTerritories  totalcases totaldeaths
1 United_States_of_America 1417889    85906      
2 Russia                    252245     2305      
3 United_Kingdom            233151    33614      
4 Spain                     229540    27321      
5 Italy                     223096    31368      
6 Brazil                    202918    13993

In [0]:
covid_sum %>%
arrange(-totaldeaths) %>%
head()

countriesAndTerritories  totalcases totaldeaths
1 United_States_of_America 1417889    85906      
2 United_Kingdom            233151    33614      
3 Italy                     223096    31368      
4 France                    141356    27425      
5 Spain                     229540    27321      
6 Brazil                    202918    13993

Next, we are interested in the monthly [incidence rate](https://en.wikipedia.org/wiki/Incidence_(epidemiology)) on a continent level. To calculate this rate, you first have to summarize the dataset on a year, month and continent level. Subsequently, you can calculate the incidence rate by using the ```mutate()``` function.

Hint 1: Some countries do not report population data. To ignore NA values you can use the parameter ```na.rm=TRUE``` in most dplyr functions.

Hint 2: You can remove the continet ```other``` prior to your analysis.

* filter out Other

In [0]:
covid %>%
filter(continentExp != "Other") -> covid_continent
covid_continent %>%
distinct(continentExp)

continentExp
1 Asia        
2 Europe      
3 Africa      
4 America     
5 Oceania

* replace NA with 0 (not sure if this worked properly)

In [0]:
covid_continent[is.na(covid_continent)] <- 0
covid_continent %>% head()

Warning message in `[<-.factor`(`*tmp*`, thisvar, value = 0):
“invalid factor level, NA generated”


dateRep    day month year cases deaths countriesAndTerritories geoId
1 20/05/2020 20  5     2020  581   5     Afghanistan             AF   
2 19/05/2020 19  5     2020  408   4     Afghanistan             AF   
3 18/05/2020 18  5     2020  262   1     Afghanistan             AF   
4 17/05/2020 17  5     2020    0   0     Afghanistan             AF   
5 16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
6 15/05/2020 15  5     2020  113   6     Afghanistan             AF   
  countryterritoryCode popData2018 continentExp
1 AFG                  37172386    Asia        
2 AFG                  37172386    Asia        
3 AFG                  37172386    Asia        
4 AFG                  37172386    Asia        
5 AFG                  37172386    Asia        
6 AFG                  37172386    Asia

* calculate Continent population

In [0]:
covid_continent %>%
distinct(popData2018, continentExp) %>%
group_by(continentExp) %>%
summarize(contpop = sum(popData2018, na.rm=TRUE)) -> covid_cont_pop
covid_cont_pop

continentExp contpop   
1 Africa       1270968396
2 America      1005667059
3 Asia         4468915667
4 Europe        764402254
5 Oceania        40152057

* compute monthly cases

In [0]:
covid_continent %>%
group_by(month, year, continentExp) %>%
summarise(monthly_cases = sum(cases)) -> covid_monthly_cases
covid_monthly_cases

month year continentExp monthly_cases
1   1    2020 Africa             0      
2   1    2020 America            9      
3   1    2020 Asia            9766      
4   1    2020 Europe            17      
5   1    2020 Oceania            7      
6   2    2020 Africa             3      
7   2    2020 America           76      
8   2    2020 Asia           73468      
9   2    2020 Europe          1126      
10  2    2020 Oceania           19      
11  3    2020 Africa          5134      
12  3    2020 America       188660      
13  3    2020 Asia           85873      
14  3    2020 Europe        426089      
15  3    2020 Oceania         5298      
16  4    2020 Africa         31598      
17  4    2020 America      1104862      
18  4    2020 Asia          333821      
19  4    2020 Europe        863874      
20  4    2020 Oceania         2812      
21  5    2020 Africa         54630      
22  5    2020 America       893300      
23  5    2020 Asia          330482      
24  5    2020 Europe        449445      
25  5    2020 Oceania          364      
26 12    2019 Africa             0      
27 12    2019 America            0      
28 12    2019 Asia              27      
29 12    2019 Europe             0      
30 12    2019 Oceania            0

* Join monthly cases with the population df. 
* Probably not the most elegant way to do this, but the only that I got it to work

In [0]:
covid_monthly_cases %>%
left_join(covid_cont_pop, covid_monthly_cases, by= "continentExp") -> covid_join
covid_join

month year continentExp monthly_cases contpop   
1   1    2020 Africa             0       1270968396
2   1    2020 America            9       1005667059
3   1    2020 Asia            9766       4468915667
4   1    2020 Europe            17        764402254
5   1    2020 Oceania            7         40152057
6   2    2020 Africa             3       1270968396
7   2    2020 America           76       1005667059
8   2    2020 Asia           73468       4468915667
9   2    2020 Europe          1126        764402254
10  2    2020 Oceania           19         40152057
11  3    2020 Africa          5134       1270968396
12  3    2020 America       188660       1005667059
13  3    2020 Asia           85873       4468915667
14  3    2020 Europe        426089        764402254
15  3    2020 Oceania         5298         40152057
16  4    2020 Africa         31598       1270968396
17  4    2020 America      1104862       1005667059
18  4    2020 Asia          333821       4468915667
19  4    2020 Europe        863874        764402254
20  4    2020 Oceania         2812         40152057
21  5    2020 Africa         54630       1270968396
22  5    2020 America       893300       1005667059
23  5    2020 Asia          330482       4468915667
24  5    2020 Europe        449445        764402254
25  5    2020 Oceania          364         40152057
26 12    2019 Africa             0       1270968396
27 12    2019 America            0       1005667059
28 12    2019 Asia              27       4468915667
29 12    2019 Europe             0        764402254
30 12    2019 Oceania            0         40152057

removing scientific annotation (e-04 etc.)

In [0]:
options(scipen = 999)

Calculating incidence rate per 100.000 rounded to three decimals

In [0]:
covid_join %>%
mutate(incidence_rate = (monthly_cases / contpop)*100000) %>%
mutate_if(is.numeric, round, 3) %>%
arrange(-month) -> covid_inc_rate
covid_inc_rate

`mutate_if()` ignored the following grouping variables:
Columns `month`, `year`



month year continentExp monthly_cases contpop    incidence_rate
1  12    2019 Africa             0       1270968396   0.000       
2  12    2019 America            0       1005667059   0.000       
3  12    2019 Asia              27       4468915667   0.001       
4  12    2019 Europe             0        764402254   0.000       
5  12    2019 Oceania            0         40152057   0.000       
6   5    2020 Africa         54630       1270968396   4.298       
7   5    2020 America       893300       1005667059  88.827       
8   5    2020 Asia          330482       4468915667   7.395       
9   5    2020 Europe        449445        764402254  58.797       
10  5    2020 Oceania          364         40152057   0.907       
11  4    2020 Africa         31598       1270968396   2.486       
12  4    2020 America      1104862       1005667059 109.864       
13  4    2020 Asia          333821       4468915667   7.470       
14  4    2020 Europe        863874        764402254 113.013       
15  4    2020 Oceania         2812         40152057   7.003       
16  3    2020 Africa          5134       1270968396   0.404       
17  3    2020 America       188660       1005667059  18.760       
18  3    2020 Asia           85873       4468915667   1.922       
19  3    2020 Europe        426089        764402254  55.741       
20  3    2020 Oceania         5298         40152057  13.195       
21  2    2020 Africa             3       1270968396   0.000       
22  2    2020 America           76       1005667059   0.008       
23  2    2020 Asia           73468       4468915667   1.644       
24  2    2020 Europe          1126        764402254   0.147       
25  2    2020 Oceania           19         40152057   0.047       
26  1    2020 Africa             0       1270968396   0.000       
27  1    2020 America            9       1005667059   0.001       
28  1    2020 Asia            9766       4468915667   0.219       
29  1    2020 Europe            17        764402254   0.002       
30  1    2020 Oceania            7         40152057   0.017

Modify your code to also calculate the death rate on the same aggregation level.

* Aggregate on month, year, continent and compute monthly cases / deaths

* compute death_rate 

In [0]:
covid_continent %>%
group_by(month, year, continentExp) %>%
summarise(monthly_cases = sum(cases), monthly_deaths = sum(deaths)) %>%
mutate(death_rate = monthly_deaths / monthly_cases) -> covid_double
covid_double

month year continentExp monthly_cases monthly_deaths death_rate
1   1    2020 Africa             0            0           NaN     
2   1    2020 America            9            0         0.000     
3   1    2020 Asia            9766          213         0.022     
4   1    2020 Europe            17            0         0.000     
5   1    2020 Oceania            7            0         0.000     
6   2    2020 Africa             3            0         0.000     
7   2    2020 America           76            0         0.000     
8   2    2020 Asia           73468         2679         0.036     
9   2    2020 Europe          1126           23         0.020     
10  2    2020 Oceania           19            0         0.000     
11  3    2020 Africa          5134          166         0.032     
12  3    2020 America       188660         3687         0.020     
13  3    2020 Asia           85873         3990         0.046     
14  3    2020 Europe        426089        27125         0.064     
15  3    2020 Oceania         5298           21         0.004     
16  4    2020 Africa         31598         1425         0.045     
17  4    2020 America      1104862        70904         0.064     
18  4    2020 Asia          333821        11321         0.034     
19  4    2020 Europe        863874       105674         0.122     
20  4    2020 Oceania         2812           95         0.034     
21  5    2020 Africa         54630         1312         0.024     
22  5    2020 America       893300        55089         0.062     
23  5    2020 Asia          330482         7214         0.022     
24  5    2020 Europe        449445        31527         0.070     
25  5    2020 Oceania          364           11         0.030     
26 12    2019 Africa             0            0           NaN     
27 12    2019 America            0            0           NaN     
28 12    2019 Asia              27            0         0.000     
29 12    2019 Europe             0            0           NaN     
30 12    2019 Oceania            0            0           NaN

## Task 2

**Contact tracing**

Governments rely on contact tracing to avoid the spread of the COVID-19 disease. Once a new case has been identified, health authorities track down people who may have been in contact with the sick person. 

Recently, there has been a new case and you are tasked to help local authorities to identify all contact persons. Initially, your only infomation is that the infected person (Mike) started to show symptoms on May 1, 2020 and was tested positiv and quarantined on the same day. However, infected persons can be contagious for up to 2 weeks without showing any symptoms. Hence, you have to reconstruct all contacts in that time window. 

The health authorities provide you with a dataframe ```df``` containing the personal information on Mike.

In [2]:
df = data.frame(first_name = 'Mike', last_name = 'Williams', job = 'Student', passport_number = 204934343, phone_number = '+1-202-555-0121', quarantine_date = '2020-05-1', stringsAsFactors = FALSE)
df

first_name last_name job     passport_number phone_number    quarantine_date
1 Mike       Williams  Student 204934343       +1-202-555-0121 2020-05-1

In order to identify all contacts you first have to find the date Mike stated to be contagious.

Hint: You should use the ```lubridate``` package.

In [3]:
df %>%
mutate(quarantine_date= ymd(quarantine_date)) 

first_name last_name job     passport_number phone_number    quarantine_date
1 Mike       Williams  Student 204934343       +1-202-555-0121 2020-05-01

During your investigations you find out that Mike visited the local gym once in the last month. Luckily, the gym uses an electronic gate and logs all visits with time stamps and member id. To help you reconstruct the data, the gym owner provides you with an additional file containing information on the members.  

You can downlaod the member data from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_data.csv
and the visitor logs from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_log.csv

Load both dataframes and identify all possible contact persons (by name and member number) that could have met Mike in the gym for at least 15 minutes while he was contagious.



Hint: You might have to use ```lubridate``` to transform the checkin and checkout columns


In [4]:
member_data <- as.data.frame(read_csv(url("https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_data.csv")))
member_data %>% head()

Parsed with column specification:
cols(
  names = col_character(),
  member_id = col_double()
)



names               member_id
1 Williams, Mike      100000   
2 Atwater III, Davyen 100001   
3 Pham, Kaili         100002   
4 el-Khalaf, Ridwaan  100003   
5 Harmon, Naomi       100004   
6 Schutte, Andrew     100005

In [7]:
visitor_logs <- as.data.frame(read_csv(url("https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_log.csv")))
visitor_logs %>% head()

Parsed with column specification:
cols(
  member_id = col_double(),
  checkin = col_datetime(format = ""),
  checkout = col_datetime(format = "")
)



member_id checkin             checkout           
1 100000    2020-04-27 20:50:08 2020-04-27 22:29:08
2 100001    2020-04-01 02:40:03 2020-04-01 03:18:03
3 100001    2020-04-01 16:01:10 2020-04-01 16:51:10
4 100001    2020-04-09 08:49:04 2020-04-09 09:19:04
5 100001    2020-04-09 20:07:06 2020-04-09 21:18:06
6 100001    2020-04-10 20:54:41 2020-04-10 22:41:41

In [10]:
visitor_logs %>%
left_join(member_data, by= "member_id") -> gym_joined
gym_joined %>%
mutate(checkin = ymd_hms(checkin), checkout = ymd_hms(checkout)) %>%
head()


member_id checkin             checkout            names              
1 100000    2020-04-27 20:50:08 2020-04-27 22:29:08 Williams, Mike     
2 100001    2020-04-01 02:40:03 2020-04-01 03:18:03 Atwater III, Davyen
3 100001    2020-04-01 16:01:10 2020-04-01 16:51:10 Atwater III, Davyen
4 100001    2020-04-09 08:49:04 2020-04-09 09:19:04 Atwater III, Davyen
5 100001    2020-04-09 20:07:06 2020-04-09 21:18:06 Atwater III, Davyen
6 100001    2020-04-10 20:54:41 2020-04-10 22:41:41 Atwater III, Davyen

David could have been contagious 2 weeks before 1.05.2020 = 17.04.2020

In [11]:
gym_joined %>%
filter(names == "Williams, Mike")

member_id checkin             checkout            names         
1 1e+05     2020-04-27 20:50:08 2020-04-27 22:29:08 Williams, Mike

only relevant timeframe is 27.04.2020 from 20:50:08 until 22:29:08 (contact for at least 15 min)

In [13]:
contagion <- interval(ymd_hms("2020-04-27 20:50:08"), ymd_hms("2020-04-27 22:29:08"))
contagion

[1] 2020-04-27 20:50:08 UTC--2020-04-27 22:29:08 UTC

In [14]:
#testing my created contagion interval
ymd_hms("2020-04-27 20:53:08") %within% contagion

[1] TRUE

Additionally, you learn that Mike took several exams in the university. While the university has social distancing measures in place it is still important to identify all students sitting close to Mike during one of the exams. Students with seat numbers in the range of -5 to +5 to Mike have to be identified. 

You can download a list of all exams Mike was registered for from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/exams_data.csv





In [0]:
# Insert your code here

After downloading and inspecting the files you have to find all exams taken by Mike while he was contagious.

In [0]:
# Insert your code here

Next, identify all students that came into contact with Mike during one of the relevant exams. Store the first and last names of the students in a new data frame.

In [0]:
# Insert your code here